sa/sa.go

```diff
@@ -2099,34 +2245,207 @@ func (ssa *SQLStorageAuthority) getChallengesImpl(db dbSelector, authID string)
+// GetAuthorization2 returns the authz2 style authorization identified by the provided ID or an error.
+// If no authorization is found matching the ID a berrors.NotFound type error is returned. This method
+// is intended to deprecate GetAuthorization.
+func (ssa *SQLStorageAuthority) GetAuthorization2(ctx context.Context, id *sapb.AuthorizationID2) (*corepb.Authorization, error) {
+	obj, err := ssa.dbMap.Get(authz2Model{}, *id.Id)
 	if err != nil {
 		return nil, err
 	}
 	if obj == nil {
-		return nil, berrors.NotFoundError("authorization %d not found", id)
+		return nil, berrors.NotFoundError("authorization %d not found", *id.Id)
 	}
 	return modelToAuthzPB(obj.(*authz2Model))
 }
 
+func authz2ModelMapToPB(m map[string]authz2Model) (*sapb.Authorizations, error) {
+	resp := &sapb.Authorizations{}
+	for k, v := range m {
+		// Make a copy of k because it will be reassigned with each loop.
+		kCopy := k
+		authzPB, err := modelToAuthzPB(&v)
+		if err != nil {
+			return nil, err
+		}
+		resp.Authz = append(resp.Authz, &sapb.Authorizations_MapElement{Domain: &kCopy, Authz: authzPB})
+	}
+	return resp, nil
+}
+
+// GetAuthorizations2 returns any valid or pending authorizations that exist for the list of domains
+// provided. If both a valid and pending authorization exist only the valid one will be returned. This
+// method is intended to deprecate GetAuthorizations.
+func (ssa *SQLStorageAuthority) GetAuthorizations2(ctx context.Context, req *sapb.GetAuthorizationsRequest) (*sapb.Authorizations, error) {
+	qmarks := make([]string, len(req.Domains))
+	for i := range req.Domains {
+		qmarks[i] = "?"
+	}
+
+	queryPrefix := fmt.Sprintf("SELECT %s FROM authz2", authz2Fields)
+	if *req.RequireV2Authzs {
+		queryPrefix += ` JOIN orderToAuthz
+			ON id = CONVERT(authzID, INTEGER)`
+	}
+	query := fmt.Sprintf(
+		`%s WHERE
+		registrationID = ? AND
+		expires > ? AND
+		status IN (?,?) AND
+		identifierValue IN (%s)`,
+		queryPrefix,
+		strings.Join(qmarks, ","),
+	)
+
+	var authzModels []authz2Model
+	params := []interface{}{
+		*req.RegistrationID,
+		time.Unix(0, *req.Now),
+		statusToUint[string(core.StatusValid)],
+		statusToUint[string(core.StatusPending)],
+	}
+	for _, n := range req.Domains {
+		params = append(params, n)
+	}
+	_, err := ssa.dbMap.Select(
+		&authzModels,
+		query,
+		params...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	authzModelMap := make(map[string]authz2Model)
+	for _, am := range authzModels {
+		if existing, present := authzModelMap[am.IdentifierValue]; !present ||
+			uintToStatus[existing.Status] == string(core.StatusPending) && uintToStatus[am.Status] == string(core.StatusValid) {
+			authzModelMap[am.IdentifierValue] = am
+		}
+	}
+
+	authzsPB, err := authz2ModelMapToPB(authzModelMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we still need more authorizations look for them in the old
+	// style authorization tables.
+	//
+	// TODO: In theory, we don't _need_ to do this. Instead we could
```

Comment:
I think we should keep this code, returning old style authorizations if needed. The reasoning: Some of our integrators rely on authorizations staying valid for 30 days, and getting reused in new orders. If we stopped reusing valid authzs when we flip the flag, we'd break those integrators. We could provide advance notice and so on, but I think it's better to keep this storage migration as transparent to our users as possible.